airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.core.util
+
+import com.google.common.annotations.VisibleForTesting
+import io.airbyte.commons.string.Strings
+import io.github.oshai.kotlinlogging.KotlinLogging
+import jakarta.inject.Singleton
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Predicate
+import java.util.stream.Collectors
+import kotlin.system.exitProcess
+import org.apache.commons.lang3.ThreadUtils
+import org.apache.commons.lang3.concurrent.BasicThreadFactory
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Collection of utility methods to ensure that any threads launched by a connector are cleaned up
+ * when an operation completes.
+ */
+@Singleton
+class ShutdownUtils {
+    companion object {
+        private const val TYPE_AND_DEDUPE_THREAD_NAME: String = "type-and-dedupe"
+        private const val MICRONAUT_SCHEDULER_THREAD_NAME: String = "scheduled-executor-thread"
+        private const val FORCED_EXIT_CODE: Int = 2
+        const val EXIT_THREAD_DELAY_MINUTES: Int = 70
+        val EXIT_HOOK: Runnable = Runnable { exitProcess(FORCED_EXIT_CODE) }
+        const val INTERRUPT_THREAD_DELAY_MINUTES: Int = 60
+
+        /**
+         * Filters threads that should not be considered when looking for orphaned threads at
+         * shutdown of the integration runner.
+         *
+         * **N.B.** Daemon threads don't block the JVM if the main `currentThread` exits, so they
+         * are not problematic. Additionally, ignore database connection pool threads, which stay
+         * active so long as the database connection pool is open.
+         */
+        @VisibleForTesting
+        val ORPHANED_THREAD_FILTER: Predicate<Thread> =
+            Predicate<Thread> { runningThread: Thread ->
+                (runningThread.name != Thread.currentThread().name &&
+                    !runningThread.isDaemon &&
+                    (TYPE_AND_DEDUPE_THREAD_NAME != runningThread.name &&
+                        !runningThread.name.startsWith(MICRONAUT_SCHEDULER_THREAD_NAME)))
+            }
+    }
```

Comment:
I'm worried that this approach isn't going to age well. Can we instead maintain a list of thread names that shouldn't be shut down and nuke anything not matching the list?

Reply:
@postamar I'll create a ticket for this.  This is a port of the existing logic.  I have tried to not touch things that are not part of the main path to getting this change in, as it is already taking weeks to get what should be a simple, net-new first pass merged.  This feels like something that the team should take on as technical debt paydown.

Reply:
Yeah let's defer this in that case. Thanks for creating the ticket.